### ISSUE TITLE
Model Support: BAAI/bge-large-en-v1.5

### ISSUE THREAD

**Issue by suryajasper** [martin-luecke/rocm-systems#103]
- [x] unmodified model runs on gfx942 OR gfx950
- [x] relevant kernels extracted (link to the hotswap-testing repository)
- [x] transpilation succeeds on gfx942 OR gfx950
- [x] individual kernels run with synthetic data
- [x] transpiled kernels run and produces result within FP noise boundary on synthetic data (no NaNs)
- [x] whole model runs with transpiled kernels
- [x] relevant throughput metric (token/s, frame/s) measured and reported below (alongside git hashes used)
- [x] all of the above on BOTH gfx942 AND gfx950

**Comment by ftynse** [comment]
Could you please post instructions on how to reproduce your runs here?

**Comment by suryajasper** [comment]
## BAAI/bge-large-en-v1.5 — Enablement Status

### Summary

BAAI/bge-large-en-v1.5 (BERT-large, 335M params, 1024-dim embeddings) is working end-to-end through HotSwap on gfx950 with numerically correct results and <5% throughput overhead.

Two transpiler bugs were found and fixed during enablement. All 5 extracted Triton kernels transpile successfully and pass standalone numeric validation, and the full model produces embeddings within float32 precision of native execution.

### Checklist

- [x] Unmodified model runs on gfx950
- [x] Relevant kernels extracted ([rocm-hotswap-testing@`1ff7fec`](https://github.com/harsh-amd/rocm-hotswap-testing/tree/users/surya/bge-large-en-v1.5-kernels))
- [x] Transpilation succeeds on gfx950 (all 5 kernels, raise_cli 22/22 variants)
- [x] Individual kernels run with synthetic data (5/5 manifests pass `test_numeric.py`)
- [x] Transpiled kernels produce correct results (max embedding diff 6e-8, cosine 1.000000)
- [x] Whole model runs with transpiled kernels (torch.compile + inductor + HotSwap)
- [x] Throughput measured: 10,405 sent/s HotSwap vs 10,941 native at batch=64 (4.9% overhead)
- [ ] Verification on gfx942 (no gfx942 hardware available in current environment)

### Transpiler fixes required

Two bugs were fixed on branch [`users/surya/BAAI-bge-large-support`](https://github.com/martin-luecke/llvm-project/tree/users/surya/BAAI-bge-large-support):

1. **`7563cab` — Fix operand swap in S_ANDN2/S_ORN2_SAVEEXEC_B32**
   The raiser computed `EXEC = EXEC & ~SRC` instead of `EXEC = SRC & ~EXEC`. This caused the "else" branch of OCML two-path approximations (tanh, erf, sin, cos) to execute with the wrong lane mask under wave32→wave64 cross-widening. All kernels using `__ocml_tanh_f32`, `__ocml_erf_f32`, `__ocml_sin_f32`, or `__ocml_cos_f32` were affected.

2. **`e21c6ac` — Seed ttmp6 dispatch state on gfx12+**
   The raiser did not initialize `ttmp6` (wave dispatch metadata) on gfx12+, producing LLVM `undef` values that could poison downstream computations.

### How to reproduce

#### Prerequisites
- gfx950 or gfx942 GPU
- ROCm 7.2+, PyTorch 2.10+, `sentence-transformers`, `transformers`
- `rocm-hotswap` wheel installed (`pip install` from `rocm-hotswap-testing/packaging/dist/`)

#### 1. Build transpiler with fixes

```bash
git clone --branch users/surya/BAAI-bge-large-support \
  https://github.com/martin-luecke/llvm-project.git llvm-project-hotswap

cd llvm-project-hotswap
cmake -S llvm -B build -G Ninja \
  -DCMAKE_BUILD_TYPE=Release \
  -DLLVM_ENABLE_PROJECTS="clang;lld" \
  -DLLVM_TARGETS_TO_BUILD="X86;AMDGPU" \
  -DLLVM_EXTERNAL_PROJECTS="device-libs;comgr" \
  -DLLVM_EXTERNAL_DEVICE_LIBS_SOURCE_DIR=$PWD/amd/device-libs \
  -DLLVM_EXTERNAL_COMGR_SOURCE_DIR=$PWD/amd/comgr \
  -DCOMGR_ENABLE_HOTSWAP_TRANSPILE=ON

cmake --build build --target amd_comgr raise_cli llc llvm-mc lld --parallel $(nproc)
```

Copy `build/lib/libamd_comgr.so*` over the wheel's bundled copy.

#### 2. Run individual kernel tests

```bash
git clone --branch users/surya/bge-large-en-v1.5-kernels \
  git@github.com:harsh-amd/rocm-hotswap-testing.git
cd rocm-hotswap-testing

# With HotSwap env configured (LD_PRELOAD, HSA_HOTSWAP_*, etc.):
eval "$(rocm-hotswap env --target-isa gfx950)"
pytest tests/test_numeric.py -k bge_triton -v
# Expected: 5 passed
```

#### 3. Run full-model E2E

```bash
rocm-hotswap run --target-isa gfx950 --source-target gfx1250:32 -- python -c "
import os, torch, numpy as np
os.environ['TRITON_CORPUS_FORCE_TARGET'] = 'gfx1250:32'
os.environ['HSA_HOTSWAP_BACKEND_ADAPTER_POLICY'] = 'triton'
from rocm_hotswap.hotswap_py.backend_adapters import apply_backend_adapters
apply_backend_adapters(source='bge_test', log_prefix='[bge]')

from transformers import AutoModel, AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained('BAAI/bge-large-en-v1.5')
model = AutoModel.from_pretrained('BAAI/bge-large-en-v1.5').cuda().eval()
model = torch.compile(model, backend='inductor')

sentences = ['The weather is lovely today.', 'It is so sunny outside!']
inputs = tokenizer(sentences, return_tensors='pt', padding=True, truncation=True).to('cuda')
with torch.no_grad(): model(**inputs)  # warmup (triggers transpilation)
with torch.no_grad(): out = model(**inputs)

emb = torch.nn.functional.normalize(out.last_hidden_state[:, 0, :], p=2, dim=1)
sim = float(emb[0] @ emb[1])
print(f'Cosine similarity (weather/sunny): {sim:.4f}')  # Expected: ~0.83
assert sim > 0.8, f'Similarity too low: {sim}'
print('PASS')
"
```

### Throughput (gfx950, sentences/second)

| Batch | Native compiled | HotSwap compiled | Overhead |
|------:|----------------:|-----------------:|---------:|
| 1 | 362 | 350 | 3.3% |
| 4 | 1,256 | 1,223 | 2.6% |
| 16 | 5,161 | 5,045 | 2.2% |
| 64 | 10,941 | 10,405 | 4.9% |

### Kernel inventory

| Kernel | Description | Standalone test |
|--------|-------------|----------------|
| `triton_per_fused_add_embedding_native_layer_norm_slice_0` | Input embedding + layer norm | PASS (8M-element zero-filled buffers; kernel hardcodes 30522×1024 embedding table offset) |
| `triton_poi_fused__scaled_dot_product_efficient_attention_...` | Attention (SDPA) post-processing | PASS |
| `triton_per_fused_add_addmm_native_layer_norm_view_2` | FFN residual add + layer norm | PASS |
| `triton_poi_fused_addmm_gelu_view_3` | FFN intermediate (GELU) | PASS |
| `triton_poi_fused_addmm_tanh_5` | Pooler (tanh) | PASS |

**Comment by ftynse** [comment]
> Two bugs were fixed on branch 

Please send isolated fixes for these as PRs. They should come with a LIT test and, ideally, a representative kernel that we can put into rocm-hotswap-testing.

> Run full-model E2E

We need a better signal than a guess of how high cosine similarity should be. Specifically, we need to see numeric differences between native and hotswap results. I'd expect something like rtol comparison of individual values in the embedding vectors.

**Comment by adedespirlet** [comment]
### BAAI/bge-large-en-v1.5
| Branch | sent/s | tok/s | HotSwap / local |
|--------|-------:|------:|----------------:|
| local | 1,340 | 12,060 | — |
| hotswap | 1,433 | 12,897 | 1.070× |

max_abs_diff: 0.016 · min_cosine_sim: 0.999972 · ✅ equiv_passed (atol=5e-2)
Translated kernels: 22 loads (run1 cold) / 10 loads (run2 warm cache), 6 unique names:
`triton_per_fused_add_embedding_native_layer_norm_slice_0`, `triton_poi_fused__scaled_dot_product_efficient_attention__to_copy_add_arange_bitwise_and_constant_pad_nd_expand_ge_index_new_ones_scalar_tensor_slice_transpose_unsqueeze_view_where_1`, `triton_per_fused_add_native_layer_norm_view_2`, `triton_poi_fused_gelu_view_3`, `triton_poi_fused__scaled_dot_product_efficient_attention__to_copy_add_arange_bitwise_and_constant_pad_nd_expand_ge_index_new_ones_scalar_tensor_slice_transpose_unsqueeze_view_where_4`, `triton_poi_fused_tanh_5`

### Git hashes
| Component | Commit |
|-----------|--------|
| `rocm-hotswap-testing` (`users/adespirl/support_embedding_models`) | `9ead193a035757f99f42247136bca462a7a4b82b` |
| `llvm-project` / HotSwap COMGR build (`users/adespirl/transpiler-fixes`) | `7e93be8e43e601a95b506c68c1eaf7832a565fdc` |
| `rocm-systems` / HotSwap ROCR adapter (`develop`) | `d96df72e314fd248e6cc7cc2770e37981236c290` |

**Comment by adedespirlet** [comment]
@suryajasper  the transpiler fixes are not yet on hotswap branch, did you want to land your fixes or do you want me to take over?

**Comment by martin-luecke** [comment]
I have a local version for the ttmp6 problem as well that I intend to open as a PR either today or Mon. 
I don't have anything for the other mentioned problem

**Comment by adedespirlet** [comment]
> I have a local version for the ttmp6 problem as well that I intend to open as a PR either today or Mon. I don't have anything for the other mentioned problem

ok great

**Comment by martin-luecke** [comment]
opened here: https://github.com/martin-luecke/llvm-project/pull/123

**Comment by tgymnich** [comment]
# PyTorch HotSwap E2E Summary: BAAI/bge-large-en-v1.5

- Run root: `/home/tgymnich/scratch/bge-large-hotswap/bge_large_en_v1_5_20260619_093831`
- Model config: `/home/tgymnich/rocm-hotswap-testing/data/pytorch/models/embedding_bge_large_en_v1_5.json`
- Workload: `hf_embedding`
- Mode: `compare`
- Source target: `gfx1250:32`
- Target GPU ISA: `gfx942`
- Adapter policy: `compile`

## local
- Passed: `True`
- Log: `/home/tgymnich/scratch/bge-large-hotswap/bge_large_en_v1_5_20260619_093831/local/run.log`
- Aggregate tok/s: `3532.852`

## hotswap
- Passed: `True`
- Log: `/home/tgymnich/scratch/bge-large-hotswap/bge_large_en_v1_5_20260619_093831/hotswap/run.log`
- Aggregate tok/s: `9650.959`
- Proof status: `translated`
- HotSwap proof: `36/36` success

## Equivalence
- Verdict: `pass`
- Status: `equivalent`
- Cases: `6` total, `0` token-divergent
